packages/react-instantsearch-core/src/widgets/InstantSearch.tsx

```diff
@@ -137,64 +129,74 @@ class InstantSearch extends Component<Props, State> {
-        onSearchStateChange: this.onSearchStateChange.bind(this),
-        onSearchParameters: this.onSearchParameters.bind(this),
-        store: this.aisManager.store,
-        widgetsManager: this.aisManager.widgetsManager,
-        mainTargetedIndex: this.props.indexName,
+        ...prevState.contextValue,
+        mainTargetedIndex: nextProps.indexName,
       },
     };
   }
 
-  componentWillReceiveProps(nextProps) {
-    // @TODO: DidUpdate
-    validateNextProps(this.props, nextProps);
-
-    if (this.props.indexName !== nextProps.indexName) {
-      this.aisManager.updateIndex(nextProps.indexName);
-      this.setState(state => ({
-        contextValue: {
-          ...state.contextValue,
-          mainTargetedIndex: nextProps.indexName,
-        },
-      }));
+  isUnmounting: boolean = false;
+  aisManager: InstantSearchManager = createInstantSearchManager({
+    indexName: this.props.indexName,
+    searchClient: this.props.searchClient,
+    initialState: this.props.searchState || {},
+    resultsState: this.props.resultsState,
+    stalledSearchDelay: this.props.stalledSearchDelay,
+  });
+  isControlled: boolean = isControlled(this.props);
+
+  state = {
+    contextValue: {
+      onInternalStateUpdate: this.onWidgetsInternalStateUpdate.bind(this),
+      createHrefForState: this.createHrefForState.bind(this),
+      onSearchForFacetValues: this.onSearchForFacetValues.bind(this),
+      onSearchStateChange: this.onSearchStateChange.bind(this),
+      onSearchParameters: this.onSearchParameters.bind(this),
+      store: this.aisManager.store,
+      widgetsManager: this.aisManager.widgetsManager,
+      mainTargetedIndex: this.props.indexName,
+    },
+  };
+
+  componentDidUpdate(prevProps: Props) {
+    const nextIsControlled = isControlled(this.props);
+
+    if (!this.isControlled && nextIsControlled) {
+      throw new Error(
+        "You can't switch <InstantSearch> from being uncontrolled to controlled"
+      );
     }
 
-    if (this.props.refresh !== nextProps.refresh) {
-      if (nextProps.refresh) {
-        this.aisManager.clearCache();
-      }
+    if (this.isControlled && !nextIsControlled) {
+      throw new Error(
+        "You can't switch <InstantSearch> from being controlled to uncontrolled"
+      );
     }
 
-    if (this.props.searchClient !== nextProps.searchClient) {
-      this.aisManager.updateClient(nextProps.searchClient);
+    /*
+     * Clear cache when `refresh` changes to `true`.
+     * Prevents users to always clear the cache on render if they forget to revert it to `false`.
+     */
+    if (this.props.refresh !== prevProps.refresh && this.props.refresh) {
+      this.aisManager.clearCache();
     }
 
     if (this.isControlled) {
-      this.aisManager.onExternalStateUpdate(nextProps.searchState);
+      this.aisManager.onExternalStateUpdate(this.props.searchState);
+    }
+
+    if (prevProps.indexName !== this.props.indexName) {
+      this.aisManager.updateIndex(this.props.indexName);
```

Comment:
This update of the `indexName` now triggers two requests. It might be because gDSP & cDU take effect into two different mico-tasks (the searches are batched into a single mico-task). The first request is triggered by the call to `updateIndex`. The second one is triggered by the cDU of one of the widgets. The widgets perform a deep-equal on the props to determine whether or not we have to schedule a new request. It was fine with the previous implementation but it's not the case anymore. The `contextValue` which provided inside the props change because of the `mainTargetedIndex`.

Reply:
I saw this, but this already present on the `v6` branch. I will try to fix this, but maybe not in this PR then.

Reply:
Ok then we can merge with the issues.